File: rlp/src/main/java/com/example/rlp/Raw.java
package com.example.rlp;

import java.util.Collection;
import java.util.Map;

import lombok.Builder;

@Builder
public class Raw<V> implements Container<V> {
    Class<V> rawType;

    Raw() {
    }

    public Raw(Class<V> rawType) {
        this.rawType = rawType;
    }

    public ContainerType getType() {
        return ContainerType.RAW;
    }

    @Override
    public Class<V> asRaw() {
        return rawType;
    }

    @Override
    public CollectionContainer<? extends Collection<V>, V> asCollection() {
        throw new RuntimeException("not a collection container");
    }

    @Override
    public MapContainer<? extends Map<?, V>, ?, V> asMap() {
        throw new RuntimeException("not a map container");
    }
}

File: rlp/src/main/java/com/example/rlp/container/Raw.kt
package com.example.rlp.container

import java.lang.RuntimeException

class Raw<V>(
    var rawType: Class<V>,
    override val type: ContainerType = ContainerType.RAW
) : Container<V> {

    override fun asRaw(): Class<V> {
        return rawType
    }

    override fun asCollection(): CollectionContainer<Collection<V>, V> {
        throw RuntimeException("not a collection container")
    }

    override fun asMap(): MapContainer<Map<*, V>, *, V> {
        throw RuntimeException("not a map container")
    }
}